django/db/models/sql/compiler.py

```diff
@@ -651,7 +651,9 @@ def find_ordering_name(self, name, opts, alias=None, default_order='ASC',
                                                        order, already_seen))
             return results
         targets, alias, _ = self.query.trim_joins(targets, joins, path)
-        return [(OrderBy(t.get_col(alias), descending=descending), False) for t in targets]
```

Comment:
To get this to work, I needed to allow `get_distinct` and `distinct_sql` to return a result, params tuple, rather than just a result, but I think this is a better approach. It's allowed me to get `DISTINCT ON` working with JSON keys. To get that, though, I had to add an `as_text` transform to get access to `#>>`, as JSON keys aren't orderable, unlike JSONB keys.

I'm still playing, and the branch is dirty, but this has got rid of my concerns above, so I think I'm on the right track.